### PR TITLE
Add acquire_value API to simplify caller code

### PR DESCRIPTION
**Pull request by franc0is**
JerryScript-DCO-1.0-Signed-off-by: François Baldassari francois@pebble.com

### CODE REVIEW

File: jerry-core/jerry.c [hunk 1]
```diff
@@ -635,6 +635,24 @@ jerry_api_release_object (jerry_api_object_t *object_p) /**< pointer acquired th
 } /* jerry_api_release_object */
 
 /**
+ * Acquire specified Jerry API value
+ */
```
Comment (LaszloLango) [inline jerry-core/jerry.c]: Missing `@return ...` comment

**Comment by LaszloLango** [comment]
Besides that missing comment the patch looks good to me.

File: jerry-core/jerry.c [hunk 2]
```diff
@@ -635,6 +635,24 @@ jerry_api_release_object (jerry_api_object_t *object_p) /**< pointer acquired th
...
+jerry_api_value_t *
+jerry_api_acquire_value (jerry_api_value_t *value_p) /**< API value */
+{
+  jerry_assert_api_available ();
+
+  if (value_p->type == JERRY_API_DATA_TYPE_STRING)
+  {
+    jerry_api_acquire_string (value_p->u.v_string);
```
Comment (LaszloLango) [inline jerry-core/jerry.c]: missing `return` statement

File: jerry-core/jerry.c [hunk 3]
```diff
@@ -635,6 +635,24 @@ jerry_api_release_object (jerry_api_object_t *object_p) /**< pointer acquired th
...
+  }
+  else if (value_p->type == JERRY_API_DATA_TYPE_OBJECT)
+  {
+    jerry_api_acquire_object (value_p->u.v_object);
```
Comment (LaszloLango) [inline jerry-core/jerry.c]: ditto

File: jerry-core/jerry.c [hunk 4]
```diff
@@ -635,6 +635,24 @@ jerry_api_release_object (jerry_api_object_t *object_p) /**< pointer acquired th
...
+  }
```
Comment (akiss77) [inline jerry-core/jerry.c]: The type of a jerry API value (given with the `jerry_api_data_type_t` enum) can be different from string or object. The function does not deal with them (there is no final `else` branch) nor does it document what the expected outcome on such a value is.

Comment (franc0is) [inline jerry-core/jerry.c]: @akiss77 In my opinion, this is an implementation detail. Perhaps in the future, new data types have acquire/release semantics as well, in which case `jerry_api_acquire_value` and `jerry_api_release_value` should do the right thing. Note that `jerry_api_release_value` does not have an else clause either.

File: jerry-core/jerry.c (hunk 2 above)
Comment (franc0is) [inline jerry-core/jerry.c]: Duh. Sorry for the silly mistake. Added the return statement at the bottom of the function.

File: jerry-core/jerry.c (hunk 1 above)
Comment (franc0is) [inline jerry-core/jerry.c]: Added.

**Comment by franc0is** [comment]
Will squash the commits after approval.

File: jerry-core/jerry.c (hunk 4 above)
Comment (akiss77) [inline jerry-core/jerry.c]: I humbly disagree. Internal-only functions can be vague about implementation details (even though it's not ideal but that's how it works in practice) but this is the _public_ API of jerry. It _may_ be a no-op for everything beyond strings and objects but it _must_ document this. (My comment above also mentioned "nor does it document".)

And true enough, `jerry_api_release_value` has a deficiency in its documentation. This is no justification for making the same mistake in a new function. Better to get the new function right _and_ fix the existing doc as well.

Comment (franc0is) [inline jerry-core/jerry.c]: Fair enough! I will add documentation to both.

**Comment by LaszloLango** [comment]
@franc0is, thank you for the updates. LGTM

**Comment by zherczeg** [comment]
LGTM.

I would like to (probably will) rename these acquire / release stuff to ref / deref.

**Comment by akiss77** [comment]
(Once rebased, we can/will land it.)

**Comment by franc0is** [comment]
rebased

**Pull request merged**